File: summary/rust/22_annealing.rs
use rand::Rng;

fn simulated_annealing<F>(initial_solution: Vec<i32>, cost_function: F) -> Vec<i32>
where
    F: Fn(&Vec<i32>) -> i32,
{
    let mut current_solution = initial_solution.clone();
    let mut current_cost = cost_function(&current_solution);
    let mut temperature = 100.0;
    let min_temperature = 0.01;
    let alpha = 0.9;
    let mut rng = rand::thread_rng();

    while temperature > min_temperature {
        let mut next_solution = current_solution.clone();
        let i = rng.gen_range(0..current_solution.len());
        let j = rng.gen_range(0..current_solution.len());
        next_solution.swap(i, j);
        let next_cost = cost_function(&next_solution);

        if next_cost < current_cost || (f64::exp((current_cost - next_cost) as f64 / temperature) > rng.gen::<f64>()) {
            current_solution = next_solution;
            current_cost = next_cost;
        }

        temperature *= alpha;
    }

    current_solution
}

// Example usage:
fn main() {
    let input_array: Vec<i32> = vec![1, 2, 3, 4, 5];
    let optimized_array = simulated_annealing(input_array, |arr| arr.iter().map(|&num| num * num).sum());
    println!("Optimized Array using Simulated Annealing: {:?}", optimized_array);
}

File: summary/python/22_annealing.py
import random
import math

def simulated_annealing(initial_solution, cost_function, temperature=100.0, cooling_rate=0.995, min_temperature=0.01):
    current_solution = initial_solution.copy()
    current_cost = cost_function(current_solution)
    best_solution = current_solution.copy()
    best_cost = current_cost

    while temperature > min_temperature:
        new_solution = current_solution.copy()
        i, j = random.sample(range(len(new_solution)), 2)
        new_solution[i], new_solution[j] = new_solution[j], new_solution[i]
        new_cost = cost_function(new_solution)

        if new_cost < current_cost or random.random() < math.exp((current_cost - new_cost) / temperature):
            current_solution = new_solution
            current_cost = new_cost

            if current_cost < best_cost:
                best_solution = current_solution.copy()
                best_cost = current_cost

        temperature *= cooling_rate

    return best_solution

def calculate_cost(arr):
    return sum(num ** 2 for num in arr)

# Example usage:
if __name__ == "__main__":
    input_array = list(map(int, input().split()))
    optimized_array = simulated_annealing(input_array, calculate_cost)
    print("Optimized Array using Simulated Annealing:", optimized_array)
